WIP: Add sync status listeners to bucket

dmsnell opened the pull request:
This patch introduces new listeners for getting the sync status of an entity.

dmsnell committed:
WIP: Add sync status listeners to bucket

This patch introduces new listeners for getting the sync status of an entity.

dmsnell committed:
Call sync object listener on all non-error updates

dmsnell committed:
Notify app of when dropping empty changes

dmsnell committed:
rc-version bump

dmsnell committed:
Bucket: Generalize listener notification pattern

In each bucket event listener notifier method (`notifyOnSaveListeners` and friends)
we have been repeating the general logic of iterating the listeners for that type
and then catching and logging exceptions thrown during the notification.

In this patch we are creating a generalized listener-iterator method `notifyListeners`
which is then used by each of the listener notifier methods so they can stick to
their specific behaviors.

While the code change count is almost zero and we've created one new method we have
moved the redundant bits about the act of iterating over the listeners into one place
where we aren't at risk of copy/paste errors. The line count would have decreased
were it not for the fact that we require three and a half lines just to setup the
`Consumer` function that performs the specific logic in each method.

`androidx.core.util.Consumer` was chosen for compatability since `java.util.Consumer`
was added in Android SDK version 24.

dmsnell committed:
rc-version bump

theck13 commented on Simperium/src/main/java/com/simperium/client/Bucket.java:
```diff
@@ -775,6 +792,22 @@ public void removeOnBeforeUpdateObjectListener(OnBeforeUpdateObjectListener<T> l
+        onSyncListeners.remove(listener);
+    }
+
+    public void addLocalQueueChangeListener(OnLocalQueueChangeListener<T> listener) {
+        onLocalQueueChangeListeners.add(listener);
+    }
+
+    public void removeLocalQueueChangeListener(OnLocalQueueChangeListener<T> listener) {
+        onLocalQueueChangeListeners.remove(listener);
+    }
```
```suggestion
    public void addOnLocalQueueChangeListener(OnLocalQueueChangeListener<T> listener) {
        onLocalQueueChangeListeners.add(listener);
    }

    public void removeOnLocalQueueChangeListener(OnLocalQueueChangeListener<T> listener) {
        onLocalQueueChangeListeners.remove(listener);
    }
```

The other methods to add and remove the listeners have "On" in the name.  We can add that to the `addLocalQueueChangeListener` and `removeLocalQueueChangeListener` methods for consistency.

theck13 commented on Simperium/src/main/java/com/simperium/client/Channel.java:
```diff
@@ -1455,6 +1470,10 @@ private void processRemoteChanges()
                     if (!remoteChange.isError() && remoteChange.isRemoveOperation()) {
                         dequeueLocalChangesForKey(remoteChange.getKey());
                     }
+
+                    if (!remoteChange.isError()) {
+                        mBucket.notifyOnSyncObjectListeners(remoteChange.getKey());
+                    }
```
```suggestion
                    if (!remoteChange.isError()) {
                        mBucket.notifyOnSyncObjectListeners(remoteChange.getKey());

                        if (remoteChange.isRemoveOperation()) {
                            dequeueLocalChangesForKey(remoteChange.getKey());
                        }                        
                    }
```

Would it make sense to merge these two `if` statements?  Do you think the nested `if`s would be less readable?

dmsnell replied:
I thought about it but I felt that the nesting would be more confusing since the notification and the dequeueing are independent things. if you don't like having both of them we could create new flags indicating the type of change

```java
boolean needToRemove = !remoteChange.isError() && removeChange.isRemoveOperation();
boolean needToNotify = !removeChange.isError();

if (needToRemove) {
	dequeueLocalChangesForKey(remoteChange.getKey());
}

if (needToNotify) {
	mBucket.notifyOnSyncObjectListeners(remoteChange.getKey());
}
```

Split into names it reads even more as independent things to me; what are your thoughts?

theck13 replied:
I'm fine with either really.  The flags may make the purpose a little clearer, but I don't think they're necessary.  I'll make this conversation as resolved.

theck13 committed:
Merge pull request #230 from Simperium/refactor/generalize-listener-notify

Bucket: Generalize listener notification pattern

dmsnell committed:
Good catch!

Co-authored-by: Tyler Heck <theck13@users.noreply.github.com>

dmsnell committed:
Name post-processing conditions at end of processRemoteChange()

At the end of `processRemoteChange()` we have been deciding whether or
not to purge remaining local changes from the queue if we receive a change
that removes the associated entity.

In this branch we've started also notifying listeners of updates to an
entity as sync events. Now, however, the logic at the end feels jumbled
and the two look related somehow.

In this commit I've given semantic names to the conditions used to
trigger each side effect in hopes that it clears up the relationship
between each one and the change, and between each other.

theck13 committed:
Add sync object and local queue change methods to bucket listener test class

theck13 committed:
Merge remote-tracking branch 'origin/add/sync-status-listeners' into add/sync-status-listeners

theck13 committed:
Update add and remove listener methods with renaming from df51296

theck13 commented:
I added the new interface methods to the `BucketListenerTest.java` file to fix the continuous integration failures.  The following commands can be run from the project's root directory while a device or emulator is connected to ensure the tests pass.

```
./gradlew connectedCheck
./gradlew cAT
```
